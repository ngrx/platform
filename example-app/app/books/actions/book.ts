import { Action } from '@ngrx/store';
import { Book } from '../models/book';

export const SEARCH = '[Book] Search';
export const SEARCH_COMPLETE = '[Book] Search Complete';
export const LOAD = '[Book] Load';
export const SELECT = '[Book] Select';

/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
export class Search implements Action {
  readonly type = SEARCH;

  constructor(public payload: string) {}
}

export class SearchComplete implements Action {
  readonly type = SEARCH_COMPLETE;

  constructor(public payload: Book[]) {}
}

export class Load implements Action {
  readonly type = LOAD;

  constructor(public payload: Book) {}
}

export class Select implements Action {
  readonly type = SELECT;

  constructor(public payload: string) {}
}

/**
 * Export a type alias of all actions in this action group
 * so that reducers can easily compose action types
 */
export type Actions = Search | SearchComplete | Load | Select;
