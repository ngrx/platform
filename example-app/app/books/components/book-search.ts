import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/distinctUntilChanged';
import { Component, Output, Input, EventEmitter } from '@angular/core';

@Component({
  selector: 'bc-book-search',
  template: `
    <md-card>
      <md-card-title>Find a Book</md-card-title>
      <md-card-content>
        <md-input-container>
          <input mdInput placeholder="Search for a book" [value]="query" (keyup)="search.emit($event.target.value)">
        </md-input-container>
        <md-spinner [class.show]="searching"></md-spinner>
      </md-card-content>
    </md-card>
  `,
  styles: [
    `
    md-card-title,
    md-card-content {
      display: flex;
      justify-content: center;
    }

    input {
      width: 300px;
    }

    md-card-spinner {
      padding-left: 60px; // Make room for the spinner
    }

    md-spinner {
      width: 30px;
      height: 30px;
      position: relative;
      top: 10px;
      left: 10px;
      opacity: 0.0;
    }

    md-spinner.show {
      opacity: 1.0;
    }
  `,
  ],
})
export class BookSearchComponent {
  @Input() query = '';
  @Input() searching = false;
  @Output() search = new EventEmitter<string>();
}
