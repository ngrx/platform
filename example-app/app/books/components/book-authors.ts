import { Component, Input } from '@angular/core';

import { Book } from '../models/book';

@Component({
  selector: 'bc-book-authors',
  template: `
    <h5 md-subheader>Written By:</h5>
    <span>
      {{ authors | bcAddCommas }}
    </span>
  `,
  styles: [
    `
    h5 {
      margin-bottom: 5px;
    }
  `,
  ],
})
export class BookAuthorsComponent {
  @Input() book: Book;

  get authors() {
    return this.book.volumeInfo.authors;
  }
}
