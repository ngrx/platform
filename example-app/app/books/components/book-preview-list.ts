import { Component, Input } from '@angular/core';
import { Book } from '../models/book';

@Component({
  selector: 'bc-book-preview-list',
  template: `
    <bc-book-preview *ngFor="let book of books" [book]="book"></bc-book-preview>
  `,
  styles: [
    `
    :host {
      display: flex;
      flex-wrap: wrap;
      justify-content: center;
    }
  `,
  ],
})
export class BookPreviewListComponent {
  @Input() books: Book[];
}
