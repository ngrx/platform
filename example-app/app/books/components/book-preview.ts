import { Component, Input } from '@angular/core';
import { Book } from '../models/book';

@Component({
  selector: 'bc-book-preview',
  template: `
    <a [routerLink]="['/books', id]">
      <mat-card>
        <mat-card-title-group>
          <img mat-card-sm-image *ngIf="thumbnail" [src]="thumbnail"/>
          <mat-card-title>{{ title | bcEllipsis:35 }}</mat-card-title>
          <mat-card-subtitle *ngIf="subtitle">{{ subtitle | bcEllipsis:40 }}</mat-card-subtitle>
        </mat-card-title-group>
        <mat-card-content>
          <p *ngIf="description">{{ description | bcEllipsis }}</p>
        </mat-card-content>
        <mat-card-footer>
          <bc-book-authors [book]="book"></bc-book-authors>
        </mat-card-footer>
      </mat-card>
    </a>
  `,
  styles: [
    `
    :host {
      display: flex;
    }

    :host a {
      display: flex;
    }

    mat-card {
      width: 400px;
      margin: 15px;
      display: flex;
      flex-flow: column;
      justify-content: space-between;
    }

    @media only screen and (max-width: 768px) {
      mat-card {
        margin: 15px 0 !important;
      }
    }
    mat-card:hover {
      box-shadow: 3px 3px 16px -2px rgba(0, 0, 0, .5);
    }
    mat-card-title {
      margin-right: 10px;
    }
    mat-card-title-group {
      margin: 0;
    }
    a {
      color: inherit;
      text-decoration: none;
    }
    img {
      width: 60px;
      min-width: 60px;
      margin-left: 5px;
    }
    mat-card-content {
      margin-top: 15px;
      margin: 15px 0 0;
    }
    span {
      display: inline-block;
      font-size: 13px;
    }
    mat-card-footer {
      padding: 0 25px 25px;
    }
  `,
  ],
})
export class BookPreviewComponent {
  @Input() book: Book;

  get id() {
    return this.book.id;
  }

  get title() {
    return this.book.volumeInfo.title;
  }

  get subtitle() {
    return this.book.volumeInfo.subtitle;
  }

  get description() {
    return this.book.volumeInfo.description;
  }

  get thumbnail(): string | boolean {
    if (this.book.volumeInfo.imageLinks) {
      return this.book.volumeInfo.imageLinks.smallThumbnail;
    }

    return false;
  }
}
