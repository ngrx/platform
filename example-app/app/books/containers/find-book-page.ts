import { Component, ChangeDetectionStrategy } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { take } from 'rxjs/operators';

import * as fromBooks from '../reducers';
import * as book from '../actions/book';
import { Book } from '../models/book';

@Component({
  selector: 'bc-find-book-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <bc-book-search [query]="searchQuery$ | async" [searching]="loading$ | async" [error]="error$ | async" (search)="search($event)"></bc-book-search>
    <bc-book-preview-list [books]="books$ | async"></bc-book-preview-list>
  `,
})
export class FindBookPageComponent {
  searchQuery$: Observable<string>;
  books$: Observable<Book[]>;
  loading$: Observable<boolean>;
  error$: Observable<string>;

  constructor(private store: Store<fromBooks.State>) {
    this.searchQuery$ = store.pipe(select(fromBooks.getSearchQuery), take(1));
    this.books$ = store.pipe(select(fromBooks.getSearchResults));
    this.loading$ = store.pipe(select(fromBooks.getSearchLoading));
    this.error$ = store.pipe(select(fromBooks.getSearchError));
  }

  search(query: string) {
    this.store.dispatch(new book.Search(query));
  }
}
