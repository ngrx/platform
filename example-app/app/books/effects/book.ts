import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { Scheduler } from 'rxjs/Scheduler';
import { async } from 'rxjs/scheduler/async';
import { empty } from 'rxjs/observable/empty';
import { of } from 'rxjs/observable/of';

import { GoogleBooksService } from '../../core/services/google-books';
import {
  BookActionTypes,
  BookActions,
  SearchComplete,
  SearchError,
  Search,
} from '../actions/book';
import { Book } from '../models/book';
import {
  debounceTime,
  map,
  switchMap,
  skip,
  takeUntil,
  catchError,
} from 'rxjs/operators';

export const SEARCH_DEBOUNCE = new InjectionToken<number>('Search Debounce');
export const SEARCH_SCHEDULER = new InjectionToken<Scheduler>(
  'Search Scheduler'
);

/**
 * Effects offer a way to isolate and easily test side-effects within your
 * application.
 *
 * If you are unfamiliar with the operators being used in these examples, please
 * check out the sources below:
 *
 * Official Docs: http://reactivex.io/rxjs/manual/overview.html#categories-of-operators
 * RxJS 5 Operators By Example: https://gist.github.com/btroncone/d6cf141d6f2c00dc6b35
 */

@Injectable()
export class BookEffects {
  @Effect()
  search$: Observable<Action> = this.actions$.pipe(
    ofType<Search>(BookActionTypes.Search),
    debounceTime(this.debounce || 300, this.scheduler || async),
    map(action => action.payload),
    switchMap(query => {
      if (query === '') {
        return empty();
      }

      const nextSearch$ = this.actions$.pipe(
        ofType(BookActionTypes.Search),
        skip(1)
      );

      return this.googleBooks
        .searchBooks(query)
        .pipe(
          takeUntil(nextSearch$),
          map((books: Book[]) => new SearchComplete(books)),
          catchError(err => of(new SearchError(err)))
        );
    })
  );

  constructor(
    private actions$: Actions,
    private googleBooks: GoogleBooksService,
    @Optional()
    @Inject(SEARCH_DEBOUNCE)
    private debounce: number,
    /**
     * You inject an optional Scheduler that will be undefined
     * in normal application usage, but its injected here so that you can mock out
     * during testing using the RxJS TestScheduler for simulating passages of time.
     */
    @Optional()
    @Inject(SEARCH_SCHEDULER)
    private scheduler: Scheduler
  ) {}
}
