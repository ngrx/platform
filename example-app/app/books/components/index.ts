import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MaterialModule } from '@angular/material';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';

import { BookAuthorsComponent } from './book-authors';
import { BookDetailComponent } from './book-detail';
import { BookPreviewComponent } from './book-preview';
import { BookPreviewListComponent } from './book-preview-list';
import { BookSearchComponent } from './book-search';

import { PipesModule } from '../../shared/pipes';

export const COMPONENTS = [
  BookAuthorsComponent,
  BookDetailComponent,
  BookPreviewComponent,
  BookPreviewListComponent,
  BookSearchComponent,
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    RouterModule,
    PipesModule,
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class ComponentsModule {}
