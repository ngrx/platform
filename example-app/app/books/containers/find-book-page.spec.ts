import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import {
  MatCardModule,
  MatInputModule,
  MatProgressSpinnerModule,
} from '@angular/material';
import { combineReducers, Store, StoreModule } from '@ngrx/store';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';

import { BookSearchComponent } from '../components/book-search';
import { BookPreviewComponent } from '../components/book-preview';
import { BookPreviewListComponent } from '../components/book-preview-list';
import { RouterTestingModule } from '@angular/router/testing';
import { EllipsisPipe } from '../../shared/pipes/ellipsis';
import { BookAuthorsComponent } from '../components/book-authors';
import { AddCommasPipe } from '../../shared/pipes/add-commas';
import { FindBookPageComponent } from './find-book-page';
import * as book from '../actions/book';
import * as fromBooks from '../reducers';

describe('Find Book Page', () => {
  let fixture: ComponentFixture<FindBookPageComponent>;
  let store: Store<fromBooks.State>;
  let instance: FindBookPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        NoopAnimationsModule,
        StoreModule.forRoot({
          books: combineReducers(fromBooks.reducers),
        }),
        RouterTestingModule,
        MatInputModule,
        MatCardModule,
        MatProgressSpinnerModule,
        ReactiveFormsModule,
      ],
      declarations: [
        FindBookPageComponent,
        BookSearchComponent,
        BookPreviewComponent,
        BookPreviewListComponent,
        BookAuthorsComponent,
        AddCommasPipe,
        EllipsisPipe,
      ],
    });

    fixture = TestBed.createComponent(FindBookPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.get(Store);

    spyOn(store, 'dispatch').and.callThrough();
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a book.Search action on search', () => {
    const $event: string = 'book name';
    const action = new book.Search($event);

    instance.search($event);

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
