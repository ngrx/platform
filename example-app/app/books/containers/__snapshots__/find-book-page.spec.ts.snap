// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Find Book Page should compile 1`] = `
<bc-find-book-page
  books$={[Function Store]}
  error$={[Function Store]}
  loading$={[Function Store]}
  searchQuery$={[Function Store]}
  store={[Function Store]}
>
  
      <bc-book-search
    _nghost-c0=""
    ng-reflect-error=""
    ng-reflect-query=""
    ng-reflect-searching="false"
  >
    
      
    <mat-card
      _ngcontent-c0=""
      class="mat-card"
    >
      
        
      <mat-card-title
        _ngcontent-c0=""
        class="mat-card-title"
      >
        Find a Book
      </mat-card-title>
      
        
      <mat-card-content
        _ngcontent-c0=""
        class="mat-card-content"
      >
        
          
        <mat-input-container
          _ngcontent-c0=""
          class="mat-input-container mat-form-field ng-tns-c3-0 mat-primary mat-form-field-type-mat-input mat-form-field-can-float mat-form-field-hide-placeholder"
        >
          <div
            class="mat-input-wrapper mat-form-field-wrapper"
          >
            <div
              class="mat-input-flex mat-form-field-flex"
            >
              
              <div
                class="mat-input-infix mat-form-field-infix"
              >
                
            
                <input
                  _ngcontent-c0=""
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control"
                  id="mat-input-0"
                  matinput=""
                  ng-reflect-placeholder="Search for a book"
                  ng-reflect-value=""
                  placeholder="Search for a book"
                />
                
          
                <span
                  class="mat-form-field-label-wrapper mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-0"
                    class="mat-form-field-label mat-input-placeholder mat-form-field-placeholder ng-tns-c3-0 mat-empty mat-form-field-empty ng-star-inserted"
                    for="mat-input-0"
                    ng-reflect-ng-switch="false"
                  >
                    
                    
                    Search for a book
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            <div
              class="mat-input-underline mat-form-field-underline"
            >
              <span
                class="mat-input-ripple mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
              ng-reflect-ng-switch="hint"
            >
              
              
              <div
                class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c3-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity: 1;"
              >
                
                <div
                  class="mat-input-hint-spacer mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-input-container>
        
          
        <mat-spinner
          _ngcontent-c0=""
          class="mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation"
          mode="indeterminate"
          ng-reflect-diameter="30"
          ng-reflect-stroke-width="3"
          role="progressbar"
          style="width: 30px; height: 30px;"
        >
          <svg
            focusable="false"
            preserveAspectRatio="xMidYMid meet"
            style="width: 30px; height: 30px;"
            viewBox="0 0 23 23"
          >
            <circle
              cx="50%"
              cy="50%"
              r="10"
              style="stroke-dasharray: 62.83185307179586px; stroke-width: 10%;"
            />
          </svg>
        </mat-spinner>
        
        
      </mat-card-content>
      
        
      
      
      <mat-card-footer
        _ngcontent-c0=""
        class="mat-card-footer"
      >
        
      </mat-card-footer>
    </mat-card>
    
    
  </bc-book-search>
      <bc-book-preview-list
    _nghost-c1=""
    ng-reflect-books=""
  >
    
      
    
    
    
  </bc-book-preview-list>
    
</bc-find-book-page>
`;
